### ISSUE TITLE
Can't modify frozen Hash error using 2.1+ in conjunction with Raven::Processor::SanitizeData

### ISSUE THREAD

**Issue by jsdalton** [getsentry/raven-ruby#592]
We experienced a critical issue today when upgrading Sentry to version 2.1.3. Calls to `capture_exception` from our Rails application started failing with "RuntimeError: can't modify frozen Hash" errors.

Here is the full stacktrace we get when reproducing the error from a console:

```
irb(main):004:0> Raven::VERSION
=> "2.1.3"
irb(main):005:0> Raven.capture_exception Exception.new('Boom!')
RuntimeError: can't modify frozen Hash
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `block in process'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `each'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `each_with_object'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `process'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:23:in `sanitize'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `block in process'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `each'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `each_with_object'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/processor/sanitizedata.rb:18:in `process'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/client.rb:66:in `block in encode'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/client.rb:66:in `each'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/client.rb:66:in `reduce'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/client.rb:66:in `encode'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/client.rb:37:in `send_event'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/instance.rb:88:in `send_event'
        from /var/lib/gems/2.1.0/gems/sentry-raven-2.1.3/lib/raven/instance.rb:131:in `capture_type'
... 11 levels...
        from /var/lib/gems/2.1.0/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:268:in `load'
        from /var/lib/gems/2.1.0/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:268:in `block in load'
        from /var/lib/gems/2.1.0/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:240:in `load_dependency'
        from /var/lib/gems/2.1.0/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:268:in `load'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/commands/rails.rb:6:in `call'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/command_wrapper.rb:38:in `call'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application.rb:191:in `block in serve'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application.rb:161:in `fork'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application.rb:161:in `serve'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application.rb:131:in `block in run'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application.rb:125:in `loop'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application.rb:125:in `run'
        from /var/lib/gems/2.1.0/gems/spring-2.0.0/lib/spring/application/boot.rb:19:in `<top (required)>'
        from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /usr/lib/ruby/2.1.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from -e:1:in `<main>'
```

## Cause

After debugging the issue further, we determined that the problem was introduced in dc6d983f413572a0871cf6e8895e0eb8e7fdb4dd, specifically here:

https://github.com/getsentry/raven-ruby/blob/dc6d983f413572a0871cf6e8895e0eb8e7fdb4dd/lib/raven/event.rb#L24

The error occurs specifically when using sentry-raven version 2.1+ with the `Raven::Processor::SanitizeData` as a processor:

```ruby
  Raven.configure do |config|
    ...
    config.processors = [Raven::Processor::SanitizeData]
    ...
  end

```

The error occurs because the `sanitize` method in the `SanitizeData` processor attempts to recursively sanitize and update the entire event data hash being sent to Sentry. When this method encounters the new `:sdk` data hash, it attempts to overwrite one of the values within it. Because the `:sdk` value itself is set here from a class constant which is frozen (i.e. `Raven::Event::SDK`):

https://github.com/getsentry/raven-ruby/blob/dc6d983f413572a0871cf6e8895e0eb8e7fdb4dd/lib/raven/event.rb#L223

 ...a `RuntimeError` is raised since it's trying to modify the hash.

We have reverted to `2.0.2` because we depend upon the sanitization processing, but hope to be able to get back to the latest version if/when this issue is resolved.

**Comment by nateberkopec** [comment]
Interesting, we never caught this because all of our test suites use all of the default processors. It looks like the UTF8 processor creates copies of the frozen Hash, so this problem is never encountered.

Why aren't you using `Raven::Processor::UTF8Conversion`?

**Comment by nateberkopec** [comment]
FWIW I do want to solve this, because `SanitizeData` should not blow up on frozen objects.

**Comment by nateberkopec** [comment]
See #594.

**Comment by jsdalton** [comment]
Thanks for the quick response @nateberkopec.

Looking back at the project, looks like raven gem was configured way back in 2013 and nobody has touched the processors configuration since. Looking at your docs under processor:

https://docs.sentry.io/clients/ruby/config/#optional-settings

I see this:

```
config.processors = [Raven::Processor::SanitizeData]
```

So it wouldn't be surprising if we just went with this value as the most straightforward path to getting things up and running, and perhaps we overlooked the defaults. It'd be nice if the documentation were clearer about which processors are or are not included by default. 😄 

Thanks again!
